File: sort_functions/sort_functions_c/hybridSort.c
#include <stdio.h>
#include <stdlib.h>


int hybridSort(int array[], int size, int threshold);
int insertionSort(int array[], int size);
int merge(int leftArray[], int leftSize, int rightArray[], int rightSize, int array[]);


int hybridSort(int array[], int size, int threshold) {
    
    if (size <= 1) {
        return 0;  // No comparisons needed for array of size 1
    }

    if (size <= threshold) {
        insertionSort(array, size);
        // In real world, we might count comparisons in insertionSort too
        return 0;
    }

    int middle = size / 2;
    int leftSize = middle;
    int rightSize = size - middle;

    // Dynamically allocate memory for left and right arrays
    int* leftArray = (int*)malloc(leftSize * sizeof(int));
    int* rightArray = (int*)malloc(rightSize * sizeof(int));

    for (int i = 0; i < leftSize; i++) {
        leftArray[i] = array[i];
    }

    for (int i = 0; i < rightSize; i++) {
        rightArray[i] = array[leftSize + i];
    }

    int leftComparisons = hybridSort(leftArray, leftSize, threshold);
    int rightComparisons = hybridSort(rightArray, rightSize, threshold);
    int mergeComparisons = merge(leftArray, leftSize, rightArray, rightSize, array);

    free(leftArray);
    free(rightArray);

    return leftComparisons + rightComparisons + mergeComparisons;
}

File: sort_functions/sort_functions_python/hybrid_sort.py
from .insertion_sort import insertion_sort
from .merge_sort import merge_sort, merge


def hybrid_sort(array: list, threshold: int) -> tuple:

    if len(array) <= 1:
        return array, 0

    # Recursively sort both halves
    if len(array) <= threshold:
        return insertion_sort(array)

    else:
        # Split the array into two halves
        middle = len(array) // 2
        left_half = array[:middle]
        right_half = array[middle:]

        left_sorted, left_comparisons = hybrid_sort(left_half, threshold)
        right_sorted, right_comparisons = hybrid_sort(right_half, threshold)

        # Merge the two halves and count the comparisons
        merged_array, merge_comparisons = merge(left_sorted, right_sorted)

        total_comparisons = left_comparisons + right_comparisons + merge_comparisons

        return (merged_array, total_comparisons)